packages/hardhat/contracts/strategy/CrabMigration.sol

```diff
@@ -130,35 +174,95 @@ contract CrabMigration is Ownable {
     function onDeferredLiquidityCheck(bytes memory encodedData) external afterInitialized {
         require(msg.sender == EULER_MAINNET, "M4");
 
-        // 1. Borrow weth
-        uint256 crabV1Balance = crabV1.balanceOf(address(this));
-        uint256 crabV1Supply = crabV1.totalSupply();
-        (address _, uint256 id, uint256 totalCollateral, uint256 totalShort) = crabV1.getVaultDetails();
-        uint256 amountEthToBorrow = totalCollateral.wmul(crabV1Balance.wdiv(crabV1Supply));
+        FlashloanCallbackData memory data = abi.decode(encodedData, (FlashloanCallbackData));
 
-        IDToken(dToken).borrow(0, amountEthToBorrow);
-        weth.withdraw(amountEthToBorrow);
-
-        // 2. mint osqth in crab v2
-        uint256 wSqueethToMint = crabV1.getWsqueethFromCrabAmount(crabV1Balance);
-        uint256 timeAtLastHedge = crabV1.timeAtLastHedge();
-        uint256 priceAtLastHedge = crabV1.priceAtLastHedge();
-        crabV2.initialize{value: amountEthToBorrow}(
-            wSqueethToMint,
-            totalCrabV1SharesMigrated,
-            timeAtLastHedge,
-            priceAtLastHedge
-        );
+        // 1. Borrow weth
+        IDToken(dToken).borrow(0, data.amountToBorrow);
+        weth.withdraw(data.amountToBorrow);
 
-        // 3. call withdraw from crab v1
-        IERC20(wPowerPerp).approve(address(crabV1), type(uint256).max);
-        crabV1.approve(address(crabV1), crabV1Balance);
-        crabV1.withdraw(crabV1Balance);
+        // 2. Callback
+        _flashCallback(data.caller, data.amountToBorrow, data.callSource, data.callData);
 
-        // 4. Repay the weth:
-        weth.deposit{value: amountEthToBorrow}();
+        // 3. Repay the weth:
+        weth.deposit{value: data.amountToBorrow}();
         weth.approve(EULER_MAINNET, type(uint256).max);
-        IDToken(dToken).repay(0, amountEthToBorrow);
+        IDToken(dToken).repay(0, data.amountToBorrow);
+    }
+
+    function _flashCallback(
+        address _initiator,
+        uint256 _amount,
+        uint8 _callSource,
+        bytes memory _calldata
+    ) internal {
+        if (FLASH_SOURCE(_callSource) == FLASH_SOURCE.BATCH_MIGRATE) {
+            // deferLiquidityCheck can be called by anyone
+            require((_initiator == owner()), "M1");
+
+            uint256 crabV1Balance = crabV1.balanceOf(address(this));
+
+            // 2. mint osqth in crab v2
+            uint256 wSqueethToMint = crabV1.getWsqueethFromCrabAmount(crabV1Balance);
+            uint256 timeAtLastHedge = crabV1.timeAtLastHedge();
+            uint256 priceAtLastHedge = crabV1.priceAtLastHedge();
+            crabV2.initialize{value: _amount}(
+                wSqueethToMint,
+                totalCrabV1SharesMigrated,
+                timeAtLastHedge,
```

Comment:
timeAtLastHedge and priceAtLastHedge get overwritten in current Crab initialize function